casper/src/main/scala/coop/rchain/casper/engine/ApproveBlockProtocol.scala

```diff
@@ -139,6 +140,14 @@ object ApproveBlockProtocol {
                 Metrics[F].incrementCounter("genesis")
               else ().pure[F]
           _ <- Log[F].info(s"APPROVAL: received block approval from $sender")
+          _ <- EventLog[F].publish(
+                shared.Event.BlockApprovalReceived(
+                  a.candidate
+                    .flatMap(_.block.map(b => PrettyPrinter.buildStringNoLimit(b.blockHash)))
+                    .getOrElse(""),
```

Comment:
That's a lot of ugly code due to protobufs and their optionality. Let's extract that 3-times duplicated chunk into a util at least? Also, most likely "<NONE>" or "<MISSING>" would be better than "".

Moreover, I'll stand by my previous claim that the events should contain raw data rathter than strings, and the toString should happen at the usage site.

Reply:
@ArturGajowy I see your point. However, I'm not convinced that events should carry huge balls of raw data. In my opinion, events should carry as little data as possible and as much data as required. I don't like the idea that we should use the protobuf auto-generated classes in our events. On the other hand, the processing of these classes requires domain logic (from `casper`) which shouldn't leak into the `EventLog` instance somewhere else. I admit, that just using `toString` wouldn't require any domain logic. But that is an awful serialization strategy.

Reply:
> as much data as required

Yes, but not in string form. `base64(byteArray)` is bigger than `byteArray`, not to mention reference sharing.

> I don't like the idea that we should use the protobuf auto-generated classes in our events. 

I'm not arguing for that, I'm simply against strings :) In that particular case I'd use an Option[Array[Byte]] to represent the hash.

> toString wouldn't require any domain logic. But that is an awful serialization strategy.

I concur :smile: 

Now that I look at it again, I'm a bit partial on all the events being defined in the `EventLog` file. Suddenly it's the case that this is a common dependency of all modules that need to publish events. On one hand it's nice to have them all in one place, OTOH, morally, each of the events is part of the modules having the domain for them.

Anyway, no matter where we put the events, if we keep them at 'small case class containing just the data they need, either using primitives or custom-event-tree-case-classes using the former', we can probably have generic serialization to whatever format we invent, including JSON.